File: backend/src/routes/brewLogs.js
const express = require("express");
const { getDb } = require("../db");

const router = express.Router();

router.get("/", async (req, res, next) => {
  try {
    const db = getDb();
    const { productId, isPublic, limit = 200 } = req.query;

    const whereClauses = [];
    const params = [];

    // If logged in, scope to user's brew logs
    if (req.user) {
      whereClauses.push("bl.userId = ?");
      params.push(req.user.id);
    }

    if (productId) {
      const parsedProductId = parseInt(productId, 10);
      if (isNaN(parsedProductId)) return res.status(400).json({ error: "Invalid productId" });
      whereClauses.push("bi.productId = ?");
      params.push(parsedProductId);
    }
    if (isPublic !== undefined) {
      whereClauses.push("bl.isPublic = ?");
      params.push(isPublic === "1" || isPublic === "true" ? 1 : 0);
    }

    const whereSql = whereClauses.length ? "WHERE " + whereClauses.join(" AND ") : "";
    const safeLimit = Math.min(parseInt(limit, 10) || 200, 500);

    const { rows } = await db.execute({
      sql: `SELECT bl.*,
        CASE WHEN bi.productId IS NOT NULL THEN p.name ELSE bi.customName END as beanName,
        CASE WHEN bi.productId IS NOT NULL THEN p.roaster ELSE bi.customRoaster END as beanRoaster,
        bi.productId as catalogProductId,
        r.name as recipeName
      FROM brew_logs bl
      LEFT JOIN bean_inventory bi ON bi.id = bl.beanInventoryId
      LEFT JOIN products p ON p.id = bi.productId
      LEFT JOIN recipes r ON r.id = bl.recipeId
      ${whereSql}
      ORDER BY bl.createdAt DESC
      LIMIT ?`,
      args: [...params, safeLimit]
    });

    res.json(rows);
  } catch (err) { next(err); }
});

router.get("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows } = await db.execute({ sql: "SELECT * FROM brew_logs WHERE id = ?", args: [id] });
    if (!rows[0]) return res.status(404).json({ error: "Brew log not found" });
    res.json(rows[0]);
  } catch (err) { next(err); }
});

router.post("/", async (req, res, next) => {
  try {
    const db = getDb();
    const { recipeId, beanInventoryId, brewerName, grinderName, grindSize,
            coffeeGrams, waterGrams, waterTempC, brewTimeSec, rating, notes, isPublic } = req.body;
    const now = new Date().toISOString();
    const userId = req.user ? req.user.id : null;

    const statements = [{
      sql: `INSERT INTO brew_logs (recipeId, beanInventoryId, brewerName, grinderName, grindSize,
        coffeeGrams, waterGrams, waterTempC, brewTimeSec, rating, notes, isPublic, userId, createdAt)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      args: [recipeId ?? null, beanInventoryId ?? null, brewerName || null, grinderName || null,
            grindSize || null, coffeeGrams ?? null, waterGrams ?? null, waterTempC ?? null,
            brewTimeSec ?? null, rating ?? null, notes || null, isPublic !== false ? 1 : 0, userId, now]
    }];

    if (beanInventoryId && coffeeGrams) {
      statements.push({
        sql: "UPDATE bean_inventory SET gramsRemaining = MAX(0, gramsRemaining - ?), updatedAt = ? WHERE id = ?",
        args: [coffeeGrams, now, beanInventoryId]
      });
    }

    const results = await db.batch(statements, "write");
    const insertResult = results[0];

    const { rows: created } = await db.execute({ sql: "SELECT * FROM brew_logs WHERE id = ?", args: [Number(insertResult.lastInsertRowid)] });
    res.status(201).json(created[0]);
  } catch (err) { next(err); }
});

router.put("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows: existingRows } = await db.execute({ sql: "SELECT * FROM brew_logs WHERE id = ?", args: [id] });
    const existing = existingRows[0];
    if (!existing) return res.status(404).json({ error: "Brew log not found" });
    if (req.user && existing.userId && existing.userId !== req.user.id) {
      return res.status(403).json({ error: "Not authorized to edit this brew log" });
    }
    const b = req.body;
    await db.execute({
      sql: `UPDATE brew_logs SET recipeId=?, beanInventoryId=?, brewerName=?, grinderName=?, grindSize=?, coffeeGrams=?,
        waterGrams=?, waterTempC=?, brewTimeSec=?, rating=?, notes=?, isPublic=? WHERE id=?`,
      args: [b.recipeId ?? existing.recipeId, b.beanInventoryId ?? existing.beanInventoryId,
            b.brewerName ?? existing.brewerName, b.grinderName ?? existing.grinderName,
            b.grindSize ?? existing.grindSize, b.coffeeGrams ?? existing.coffeeGrams,
            b.waterGrams ?? existing.waterGrams, b.waterTempC ?? existing.waterTempC,
            b.brewTimeSec ?? existing.brewTimeSec, b.rating ?? existing.rating,
            b.notes ?? existing.notes,
            b.isPublic !== undefined ? (b.isPublic ? 1 : 0) : existing.isPublic, id]
    });
    const { rows: updated } = await db.execute({ sql: "SELECT * FROM brew_logs WHERE id = ?", args: [id] });
    res.json(updated[0]);
  } catch (err) { next(err); }
});

router.delete("/:id", async (req, res, next) => {
  try {
    const db = getDb();
    const id = parseInt(req.params.id, 10);
    if (!Number.isInteger(id)) return res.status(400).json({ error: "Invalid id" });
    const { rows: existingRows } = await db.execute({ sql: "SELECT userId FROM brew_logs WHERE id = ?", args: [id] });
    if (!existingRows[0]) return res.status(404).json({ error: "Brew log not found" });
    if (req.user && existingRows[0].userId && existingRows[0].userId !== req.user.id) {
      return res.status(403).json({ error: "Not authorized to delete this brew log" });
    }
    await db.execute({ sql: "DELETE FROM brew_logs WHERE id = ?", args: [id] });
    res.status(204).send();
  } catch (err) { next(err); }
});

module.exports = router;

File: backend-py/app/routers/brew_logs.py
"""Brew logs router — session history with inventory decrement."""

from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel

from ..database import get_db
from ..auth.dependencies import AuthUser, optional_auth

router = APIRouter(prefix="/api/brew-logs", tags=["brew-logs"])


def _rows_to_dicts(cursor):
    columns = [desc[0] for desc in cursor.description]
    return [dict(zip(columns, r)) for r in cursor.fetchall()]


def _row_to_dict(cursor, row):
    columns = [desc[0] for desc in cursor.description]
    return dict(zip(columns, row))


@router.get("/")
async def list_brew_logs(
    productId: Optional[int] = None,
    isPublic: Optional[str] = None,
    limit: int = 200,
    user: AuthUser | None = Depends(optional_auth),
):
    db = get_db()
    where_clauses = []
    params = []

    if user:
        where_clauses.append("bl.userId = ?")
        params.append(user.id)
    else:
        # Unauthenticated callers may only see public logs
        where_clauses.append("bl.isPublic = 1")

    if productId:
        where_clauses.append("bi.productId = ?")
        params.append(productId)

    if isPublic is not None:
        where_clauses.append("bl.isPublic = ?")
        params.append(1 if isPublic in ("1", "true") else 0)

    where_sql = f"WHERE {' AND '.join(where_clauses)}" if where_clauses else ""
    safe_limit = min(int(limit), 500)

    cursor = db.execute(
        f"""SELECT bl.*,
            CASE WHEN bi.productId IS NOT NULL THEN p.name ELSE bi.customName END as beanName,
            CASE WHEN bi.productId IS NOT NULL THEN p.roaster ELSE bi.customRoaster END as beanRoaster,
            bi.productId as catalogProductId,
            r.name as recipeName
        FROM brew_logs bl
        LEFT JOIN bean_inventory bi ON bi.id = bl.beanInventoryId
        LEFT JOIN products p ON p.id = bi.productId
        LEFT JOIN recipes r ON r.id = bl.recipeId
        {where_sql}
        ORDER BY bl.createdAt DESC
        LIMIT ?""",
        [*params, safe_limit],
    )
    return _rows_to_dicts(cursor)


@router.get("/{log_id}")
async def get_brew_log(log_id: int):
    db = get_db()
    cursor = db.execute("SELECT * FROM brew_logs WHERE id = ?", [log_id])
    row = cursor.fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Brew log not found")
    return _row_to_dict(cursor, row)


class BrewLogCreate(BaseModel):
    recipeId: Optional[int] = None
    beanInventoryId: Optional[int] = None
    brewerName: Optional[str] = None
    grinderName: Optional[str] = None
    grindSize: Optional[str] = None
    coffeeGrams: Optional[float] = None
    waterGrams: Optional[float] = None
    waterTempC: Optional[int] = None
    brewTimeSec: Optional[int] = None
    rating: Optional[int] = None
    notes: Optional[str] = None
    isPublic: Optional[bool] = True


@router.post("/", status_code=201)
async def create_brew_log(body: BrewLogCreate, user: AuthUser | None = Depends(optional_auth)):
    db = get_db()
    now = datetime.now(timezone.utc).isoformat()
    user_id = user.id if user else None

    cursor = db.execute(
        """INSERT INTO brew_logs (recipeId, beanInventoryId, brewerName, grinderName, grindSize,
            coffeeGrams, waterGrams, waterTempC, brewTimeSec, rating, notes, isPublic, userId, createdAt)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        [
            body.recipeId, body.beanInventoryId, body.brewerName, body.grinderName,
            body.grindSize, body.coffeeGrams, body.waterGrams, body.waterTempC,
            body.brewTimeSec, body.rating, body.notes,
            1 if body.isPublic is not False else 0, user_id, now,
        ],
    )

    # Decrement inventory if bean + grams provided — only for the owning user
    if body.beanInventoryId and body.coffeeGrams:
        ownership_clause = "AND userId = ?" if user else "AND userId IS NULL"
        ownership_params = [user.id] if user else []
        db.execute(
            f"UPDATE bean_inventory SET gramsRemaining = MAX(0, gramsRemaining - ?), updatedAt = ? WHERE id = ? {ownership_clause}",
            [body.coffeeGrams, now, body.beanInventoryId, *ownership_params],
        )

    db.commit()
    new_id = cursor.lastrowid

    cursor2 = db.execute("SELECT * FROM brew_logs WHERE id = ?", [new_id])
    row = cursor2.fetchone()
    return _row_to_dict(cursor2, row)


@router.put("/{log_id}")
async def update_brew_log(log_id: int, body: dict, user: AuthUser | None = Depends(optional_auth)):
    db = get_db()
    existing_cursor = db.execute("SELECT * FROM brew_logs WHERE id = ?", [log_id])
    existing_row = existing_cursor.fetchone()
    if not existing_row:
        raise HTTPException(status_code=404, detail="Brew log not found")

    existing = _row_to_dict(existing_cursor, existing_row)

    if existing.get("userId") and (not user or existing["userId"] != user.id):
        raise HTTPException(status_code=403, detail="Not authorized to edit this brew log")

    ALLOWED = {"recipeId", "beanInventoryId", "brewerName", "grinderName", "grindSize",
               "coffeeGrams", "waterGrams", "waterTempC", "brewTimeSec", "rating", "notes"}
    updates = {k: v for k, v in body.items() if k in ALLOWED}
    if "isPublic" in body:
        updates["isPublic"] = 1 if body["isPublic"] else 0
    if not updates:
        return existing

    set_clause = ", ".join(f"{k}=?" for k in updates)
    db.execute(
        f"UPDATE brew_logs SET {set_clause} WHERE id=?",
        [*updates.values(), log_id],
    )
    db.commit()

    cursor = db.execute("SELECT * FROM brew_logs WHERE id = ?", [log_id])
    row = cursor.fetchone()
    return _row_to_dict(cursor, row)


@router.delete("/{log_id}", status_code=204)
async def delete_brew_log(log_id: int, user: AuthUser | None = Depends(optional_auth)):
    db = get_db()
    row = db.execute("SELECT userId FROM brew_logs WHERE id = ?", [log_id]).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Brew log not found")
    if row[0] and (not user or row[0] != user.id):
        raise HTTPException(status_code=403, detail="Not authorized to delete this brew log")

    db.execute("DELETE FROM brew_logs WHERE id = ?", [log_id])
    db.commit()
    return Response(status_code=204)
